rust/connlib/tunnel/src/device_channel/tun_linux.rs

```diff
@@ -350,18 +346,95 @@ impl ioctl::Request<SetTunFlagsPayload> {
     }
 }
 
+/// Back up `/etc/resolve.conf` and then modify it in-place
+async fn configure_resolv_conf(
+    dns_config: &[IpAddr],
+    resolv_path: &Path,
+    backup_path: &Path,
+) -> Result<()> {
+    if dns_config.is_empty() {
+        tracing::info!("`dns_config` is empty, leaving `/etc/resolv.conf` unchanged");
+        return Ok(());
+    }
+
+    let text = tokio::fs::read_to_string(resolv_path)
+        .await
+        .map_err(Error::ReadResolvConf)?;
+    let parsed = resolv_conf::Config::parse(&text).map_err(|_| Error::ParseResolvConf)?;
+
+    // Back up the original resolv.conf. If there's already a backup, don't modify it
+    match tokio::fs::File::options()
+        .write(true)
+        .create_new(true)
+        .open(backup_path)
+        .await
+    {
+        Err(error) if error.kind() == std::io::ErrorKind::AlreadyExists => {
+            tracing::info!(?backup_path, "Backup path already exists, won't overwrite");
+        }
+        Err(error) => return Err(Error::WriteResolvConfBackup(error)),
+        // TODO: Would do a rename-into-place here if the contents of the file mattered more
+        Ok(mut f) => f.write_all(text.as_bytes()).await?,
+    }
```

Comment:
I think extracting this into a function could make things a bit clearer because you would be able to `?` it in here (first branch in the function would return `Ok(())`)